modules/auxiliary/server/ldap_server.rb

```diff
@@ -0,0 +1,103 @@
+          native Rex LDAP server functionality created during log4shell
+          development.
+        },
+        'Author' => [
+          'RageLtMan <rageltman[at]sempervictus>', # infrastructure
+          'Spencer McIntyre' # syntactically sane/correct Ruby LDAP object actions from early effort on l4j2 scanner
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [],
+        'Actions' => [
+          [ 'Service', { 'Description' => 'Run LDAP server' } ]
+        ],
+        'PassiveActions' => [
+          'Service'
+        ],
+        'Notes' => {
+          'Stability' => [],
+          'Reliability' => [],
+          'SideEffects' => []
+        }
+      )
+    )
+  end
+
+  #
+  # Wrapper for service execution and cleanup
+  #
+  def run
+    start_service
+    service.wait
+  rescue Rex::BindFailed => e
+    print_error "Failed to bind to port #{datastore['SRVPORT']}: #{e.message}"
+  ensure
+    stop_service
+  end
+
+  #
+  # Creates Proc to handle incoming requests
+  #
+  def on_dispatch_request(client, data)
+    return if data.strip.empty?
+
+    data.extend(Net::BER::Extensions::String)
+    begin
+      pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
+      vprint_status("LDAP request data remaining: #{data}") if !data.empty?
+      resp = case pdu.app_tag
+             when Net::LDAP::PDU::BindRequest # bind request
+               # vprint_good("Received LDAP bind request from #{client} - #{pp pdu}")
+               client.authenticated = true
+               service.encode_ldap_response(
+                 pdu.message_id,
+                 Net::LDAP::ResultCodeSuccess,
+                 '',
+                 '',
+                 Net::LDAP::PDU::BindResult
+               )
+             when Net::LDAP::PDU::SearchRequest # search request
+               # vprint_good("Received LDAP search request from #{client} - #{pp pdu}")
+               if client.authenticated || datastore['LDAP_AUTH_BYPASS']
+                 # Perform query against some loaded LDIF structure
+                 filter = Net::LDAP::Filter.parse_ldap_filter(pdu.search_parameters[:filter])
+                 attrs = pdu.search_parameters[:attributes].empty? ? :all : pdu.search_parameters[:attributes]
+                 res = service.search_ldif(filter, pdu.message_id, attrs)
+                 if res.nil? || res.empty?
+                   # vprint_status("No LDAP search results for #{filter}/#{attrs} in #{self.service.ldif}")
+                   service.encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeNoSuchObject, '', 'No such object', 5)
+                 else
+                   # vprint_status("LDAP search results for #{client} - #{pp pdu}: #{pp res}")
+                   # Send the results and return success message for callback completion
+                   cli.write(res)
+                   service.encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeSuccess, '', 'Search success', 5)
```

Comment:
~~So I can't successfully test this particular branch. One issue I'm confident is that `cli` should be `client`. After that, my test query as executed using `ldapsearch -x -H ldap://192.168.159.128 -b "dc=metasploit,dc=com"`.~~

~~First no results are found for some reason because `ldapsearch` is using `objectclass` while in the LDIF file you provide as an example, it's defined as `objectClass`. If I change the case in the LDIF file, and step through it using a debugger, I can see that results are found and this branch is taken however the response is not read by `ldapsearch`.~~

Edit: I was able to track down the issue and I've revised my suggestion to fix it. The problem was `res` is an array and needs to be treated as such. With this change in place I was able to test querying fields.

```suggestion
                   client.write(res.join)
                   service.encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeSuccess, '', 'Search success', 5)
```

Reply:
Done